### PR TITLE
[go_router] Refactored RouteMatchList and imperative APIs

### PR DESCRIPTION
**Pull request by chunhtai**
This pr refactor RouteMatchList to be a tree structure.

Added a common base class RouteMatchBase. It is extended by both RouteMatch and ShellRouteMatch.

The RouteMatch is for GoRoute, and is always a leaf node

The ShellRouteMatch is for ShellRouteBase, and is always and intermediate node with a list of child RouteMatchBase[s].

This pr also redo how push is processed. Will add a doc explain this shortly.

This is a breaking change, will write a migration guide soon.

fixes https://github.com/flutter/flutter/issues/134524
fixes https://github.com/flutter/flutter/issues/130406
fixes https://github.com/flutter/flutter/issues/126365
fixes https://github.com/flutter/flutter/issues/125752
fixes https://github.com/flutter/flutter/issues/120791
fixes https://github.com/flutter/flutter/issues/120665
fixes https://github.com/flutter/flutter/issues/113001
fixes https://github.com/flutter/flutter/issues/110512

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/lib/src/builder.dart [hunk 1]
```diff
@@ -102,339 +105,192 @@ class RouteBuilder {
...
+    required this.matchList,
+    required this.matches,
+    required this.configuration,
+    required this.errorBuilder,
+    required this.errorPageBuilder,
+  });
 
-  /// Returns the top-level pages instead of the root navigator. Used for
-  /// testing.
-  List<Page<Object?>> _buildPages(
```
Comment (chunhtai) [inline packages/go_router/lib/src/builder.dart:154]: most of the recursive build pages logic is moved to _CustomNavigatorState

File: packages/go_router/lib/src/delegate.dart [hunk 2]
```diff
@@ -73,25 +79,36 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
 
   /// Returns `true` if the active Navigator can pop.
   bool canPop() {
-    final _NavigatorStateIterator iterator = _createNavigatorStateIterator();
```
Comment (chunhtai) [inline packages/go_router/lib/src/delegate.dart:76]: previously we need to scan the list and use parent navigator key to jump around the list. Therefore i created this iterator to abstract away the logic.

After the refactor we just need to look at the last node in the RouteMatchList and walk up the tree branch.

File: packages/go_router/lib/src/match.dart [hunk 3]
```diff
@@ -16,77 +16,281 @@ import 'logging.dart';
...
+  GoRouterState buildState(
+      RouteConfiguration configuration, RouteMatchList matches);
 
   /// Generate a [RouteMatch] object by matching the `route` with
   /// `remainingLocation`.
   ///
   /// The extracted path parameters, as the result of the matching, are stored
   /// into `pathParameters`.
-  static RouteMatch? match({
+  static List<RouteMatchBase> match({
```
Comment (chunhtai) [inline packages/go_router/lib/src/match.dart:62]: This is the most complex part in this pr, previously `match` just produce a RouteMatch for the input route.

After this change, this `match` will recursively match the subroute in `route` and return the entire route matches if it finds the full match.

File: packages/go_router/lib/src/match.dart [hunk 4]
```diff
@@ -294,24 +667,110 @@ class RouteMatchList {
...
+  /// ShellRouteMatch.
+  ///
+  /// This method returns a new list as long as the target is found in the
+  /// matches' subtree.
+  ///
+  /// If a target is found, the target and every node after the target in tree
+  /// order is removed.
+  static List<RouteMatchBase> _removeRouteMatchFromList(
+      List<RouteMatchBase> matches, RouteMatchBase target) {
+    // Remove is usually caused by pop; therefore, start searching from the end.
```
Comment (hangyujin) [inline packages/go_router/lib/src/match.dart:690]: >> Remove is usually caused by pop

by `usually` do you mean there are other cases?

this function looks like it assumes all removal is pop. it not only removes the target, but also remove all matches after the target.

File: packages/go_router/lib/src/match.dart [hunk 5]
```diff
@@ -102,6 +312,116 @@ class RouteMatch {
...
+    required this.matches,
+    required this.matchedLocation,
+    required this.pageKey,
+    required this.navigatorKey,
+  }) : assert(matches.isNotEmpty);
+
+  @override
+  final ShellRouteBase route;
+
+  RouteMatch get _leafMatch {
```
Comment (hangyujin) [inline packages/go_router/lib/src/match.dart:350]: I found the name `leafMatch` a bit confusing cause a tree can have many leaves. Maybe `lastLeaf` is more clear?

File: packages/go_router/lib/src/builder.dart [hunk 6]
```diff
@@ -102,339 +105,192 @@ class RouteBuilder {
...
-
-        // Build the Page for this route
-        page = _buildPageForShellRoute(
-            context, state, match, route, pagePopContext, shellRouteContext);
-        // Place the ShellRoute's Page onto the list for the parent navigator.
-        keyToPages[routeNavKey]!.insert(shellPageIdx, page);
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    // Return a HeroController based on the app type.
```
Comment (hangyujin) [inline packages/go_router/lib/src/builder.dart:170]: nit: returns

File: packages/go_router/lib/src/match.dart [hunk 7]
```diff
@@ -16,77 +16,287 @@ import 'logging.dart';
...
+        pathParameters: pathParameters,
+        uri: uri,
+        scopedNavigatorKey: scopedNavigatorKey,
       );
+      if (subRouteMatches.isNotEmpty) {
+        break;
+      }
+    }
+    if (subRouteMatches?.isEmpty ?? true) {
+      // If not finding a sub route match, it is consider not matched for this
```
Comment (hangyujin) [inline packages/go_router/lib/src/match.dart:257]: nit: considered

File: packages/go_router/lib/src/builder.dart [hunk 8]
```diff
@@ -102,339 +105,197 @@ class RouteBuilder {
...
+          navigatorKey: configuration.navigatorKey,
+          observers: observers,
+          navigatorRestorationId: restorationScopeId,
+          onPopPageWithRouteMatch: onPopPageWithRouteMatch,
+          matchList: matchList,
+          matches: matchList.matches,
+          configuration: configuration,
+          errorBuilder: errorBuilder,
+          errorPageBuilder: errorPageBuilder,
+        ));
```
Comment (hangyujin) [inline packages/go_router/lib/src/builder.dart:121]: nit: comma

File: packages/go_router/lib/src/builder.dart [hunk 9]
```diff
@@ -102,339 +105,197 @@ class RouteBuilder {
...
-  }
+class _CustomNavigator extends StatefulWidget {
+  const _CustomNavigator({
+    super.key,
+    required this.navigatorKey,
+    required this.observers,
+    required this.navigatorRestorationId,
+    required this.onPopPageWithRouteMatch,
+    required this.matchList,
+    required this.matches,
```
Comment (hangyujin) [inline packages/go_router/lib/src/builder.dart:133]: why pass both matchList and matches?

File: packages/go_router/lib/src/match.dart [hunk 10]
```diff
@@ -257,32 +592,70 @@ class RouteMatchList {
...
+
+  static RouteMatchBase _cloneBranchAndInsertImperativeMatch(
+      RouteMatchBase branch, ImperativeRouteMatch match) {
+    if (branch is ShellRouteMatch) {
+      return branch.copyWith(
+        matches: <RouteMatchBase>[
+          _cloneBranchAndInsertImperativeMatch(branch.matches.last, match),
+        ],
+      );
+    }
```
[the remark this answers is not in the document]
Comment (hangyujin) [reply packages/go_router/lib/src/match.dart:641]: thanks, now i know

File: packages/go_router/CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,10 @@
+## 13.0.0
+
+- Refactored `RouteMatchList` and imperative APIs.
```
Comment (hangyujin) [inline packages/go_router/CHANGELOG.md:3]: nit: Refactors

File: packages/go_router/lib/src/match.dart [hunk 12]
```diff
@@ -16,77 +16,287 @@ import 'logging.dart';
...
+    required ShellRouteBase route,
+    required String matchedPath, // e.g. /family/:fid
+    required String remainingLocation, // e.g. person/p1
+    required String matchedLocation, // e.g. /family/f2
+    required Map<String, String> pathParameters,
+    required GlobalKey<NavigatorState> scopedNavigatorKey,
+    required Uri uri,
+  }) {
+    final GlobalKey<NavigatorState>? parentKey =
+        route.parentNavigatorKey == null ||
```
[the remark this answers is not in the document]
Comment (hangyujin) [reply packages/go_router/lib/src/match.dart:147]: I mean ```final GlobalKey<NavigatorState>? parentKey =
        route.parentNavigatorKey == null ||
                route.parentNavigatorKey == scopedNavigatorKey
            ? null
            : route.parentNavigatorKey;```

is same as 
```final GlobalKey<NavigatorState>? parentKey =
                route.parentNavigatorKey == scopedNavigatorKey
            ? null
            : route.parentNavigatorKey;```

[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/go_router/lib/src/match.dart:147]: oh i see, yes the logic can be reduced. Nice catch

**Comment by chunhtai** [comment]
cc @ValentinVignal

**Review by hangyujin — APPROVED** [review]
LGTM

**Pull request merged**